Survival analysis case study (#63)
tidymodels/tidymodels.org: learn/statistics/survival-case-study/index.qmd

toc: true
toc-depth: 2
include-after-body: ../../../resources.html
---

```{r}
#| label: "setup"
#| include: false
#| message: false
#| warning: false
source(here::here("common.R"))
```

```{r}
#| label: "load"
#| include: false
library(tidymodels)
library(sessioninfo)
pkgs <- c("tidymodels", "censored")
theme_set(theme_bw() + theme(legend.position = "top"))
```

## Introduction

`r article_req_pkgs(pkgs)`

Survival analysis is a field of statistics and machine learning for analyzing the time to an event. While it has its roots in medical research, the event of interest can be anything from customer churn to machine failure. Methods from survival analysis take into account that some observations may not yet have experienced the event of interest and are thus _censored_. 

Here we want to predict the time it takes for a complaint to be dispositioned^[In this context, the term _disposition_ means that there has been a decision or resolution regarding the complaint that is the conclusion of the process.] by the Department of Buildings in New York City. We are going to walk through a complete analysis from beginning to end, showing how to analyze time-to-event data.

> Let's start with loading the censored package, the parnsip extension for survival analysis models.

Reviewer: ```suggestion
Let's start with loading the tidymodels and censored packages (the parsnip extension package for survival analysis models).
```

```{r}
library(tidymodels)
library(censored)
```

## The buildings complaints data

> The city of New York publishes data on the [complaints](https://data.cityofnewyork.us/Housing-Development/DOB-Complaints-Received/eabe-havv/about_data) received by the Department of Buildings. The data includes information on the type of complaint, the date it was entered in their records, the date it was dispositioned, and the location of the building the complaint was about. We are using a subset of the data, available in the modeldatatoo package.

Reviewer: We need the dev version of modeldatatoo. Will we send that in before this is merged? Also, we might want to note the version requirement for modeldatatoo.

Author: Good catch! @EmilHvitfeldt would you do a release of modeldatatoo, please? 🙏

Reviewer 2: Yes, I'll get to that sometime today

Author: Thank you! 🙌

```{r}
data(building_complaints, package = "modeldatatoo")
glimpse(building_complaints)
```

Before we dive into survival analysis, let's get a impression of how the complaints are distributed across the city. We have complaints in all five boroughs, albeit with a somewhat lower density of complaints in Staten Island.

```{r nyc-map}
#| out.width: "100%"
#| echo: false
#| fig.cap: "Building complaints in New York City"
#| fig.alt: "A scatter plot of building complaint locations superimposed on a street map. Complaints are raised in all five boroughs." 

knitr::include_graphics("nyc_building_complaints.png")
```

> In the dataset, we can see the `days_to_disposition` as well as the `status` of the complaint. For a complaint with status `"ACTIVE"`, the time to disposition is censored, meaning we do know that is has taken at least that long, but not how long it will have taken eventually. 

Reviewer: ```suggestion
In the dataset, we can see the `days_to_disposition` as well as the `status` of the complaint. For a complaint with the status "ACTIVE", the time to disposition is censored, meaning we do know that it has taken at least that long, but not how long for it to be completely resolved. 
```

The standard form for time-to-event data are `Surv` objects which capture the time as well as the event status. As with all transformations of the response, it is advisable to do this before heading into the model fitting process with tidymodels.

```{r}
building_complaints <- building_complaints %>% 
  mutate(
    disposition_surv = Surv(days_to_disposition, status == "CLOSED"), 
    .keep = "unused"
  )
```

## Data splitting and resampling

> For our resampling strategy, let's use a 3-way split into training, validation, and test set.

Reviewer: ```suggestion
For our resampling strategy, let's use a [3-way split](https://www.tmwr.org/resampling#validation) into training, validation, and test set.
```

```{r}
set.seed(403)
complaints_split <- initial_validation_split(building_complaints)
```

> First, let's pull out the training data and have a brief look at the response.

Reviewer: ```suggestion
First, let's pull out the training data and have a brief look at the response using a [Kaplan-Meier curve](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3059453/). 
```

```{r}
spec_survreg <- survival_reg() %>% 
  set_engine("survival") %>% 
  set_mode("censored regression")
```

We have several missing values in `complaint_priority` that we are turning into a separate category, `"unknown"`. We are also combining the less common categories for `community_board` and `unit` into an `"other"` category to reduce the number of levels in the predictors. The complaint category often does not tell us much more than the unit, with several complaint categories being handled by a specific unit only. This can lead to the model being unable to estimate some of the coefficients. Since our goal here is only to get a rough idea of how well the model performs, we are removing the complaint category for now.

```{r}
rec_other <- recipe(disposition_surv ~ ., data = complaints_train) %>% 
  step_unknown(complaint_priority) %>% 

>   step_rm(complaint_category) %>% 

Reviewer: Maybe change the role instead? That would keep it in the data set for _post hoc_ analysis.

Author: That's a cool thing but I'd keep it as is because it's not directly related to conducting the analysis as is. We could add a section where we do this kind of post hoc analysis but that strikes me more as "cool things tidymodels can do" rather than the main goal here of "here's how tidymodels works for survival analysis".

  step_novel(community_board, unit) %>%
  step_other(community_board, unit, threshold = 0.02)
```

We combine the recipe and the model into a workflow. This allows us to easily resample the model because all preprocessing steps are applied to the training set and the validation set for us.

```{r}
wflow_survreg <- workflow() %>% 
  add_recipe(rec_other) %>% 
  add_model(spec_survreg)
```

> Let's access the resampling set from the initial split and fit the model. We are calculating several different performance metrics here: the Brier score, its integrated version, the area under the ROC curve, and the concordance index. Note that all of these are used in a version tailored to survival analysis. The concordance index uses the predicted event time to measure the model's ability to rank the observations correctly. The Brier score and the ROC curve use the predicted probability of survival at a given time point. We are evaluating these metrics every 30 days up to 300 days, as provided in the `eval_time` argument. The Brier score is a measure of the accuracy of the predicted probabilities while the ROC curve is a measure of the model's ability to discriminate between events and non-events at the given time point. Because these metrics are defined "at a given time point" they are also referred to as dynamic metrics. 

Reviewer: ```suggestion
Let’s access the training and validation sets from the initial split and fit the model. We are calculating several performance metrics: the Brier score, its integrated version, the area under the ROC curve, and the concordance index. Note that all of these are used in a version tailored to survival analysis. The concordance index uses the predicted event time to measure the model’s ability to rank the observations correctly. The Brier score and the ROC curve use the predicted probability of survival at a given time. We evaluate these metrics every 30 days up to 300 days, as provided in the `eval_time` argument. The Brier score is a measure of the accuracy of the predicted probabilities, while the ROC curve is a measure of the model’s ability to discriminate between events and non-events at the given time point. Because these metrics are defined “at a given time,” they are also referred to as *dynamic metrics*.
```

::: {.callout-tip}
For more information see the [Dynamic Performance Metrics for Event Time Data](../survival-metrics/) article.
:::

```{r}

> complaints_rset <- validation_set(complaints_split)

Reviewer: ```suggestion
# validation_set() extracts the training and validation sets
# in a manner that emulates a single resample of the data. 
# This enables fit_resamples() and other functions in the 
# tune package to be used as if it were some other resampling
# scheme (such as cross-validation). 

complaints_rset <- validation_set(complaints_split)
```

OR maybe put that paragraph in text instead of code comments

set.seed(1)
res_survreg <- fit_resamples(
  wflow_survreg,
  complaints_rset,
  metrics = metric_set(brier_survival_integrated, brier_survival, roc_auc_survival,
                       concordance_survival),
  eval_time = seq(0, 300, 30), 
  control = control_resamples(save_workflow = TRUE, save_pred = TRUE)
)
```

> So how do predictions from censored regression models look like?

Reviewer: ```suggestion
The structure of survival model predictions is slightly different from classification and regression model predictions:
```

```{r}
preds <- collect_predictions(res_survreg)
preds
```

The predicted survival time is in the `.pred_time` column and the predicted survival probabilities are in the `.pred` list column. 

```{r}
preds$.pred[[6]]
```

For each observation, `.pred` contains a tibble with the evaluation time `.eval_time` and the corresponding survival probability `.pred_survival`. The column `.weight_censored` contains the weights used in the calculation of the dynamic performance metrics. 

::: {.callout-tip}
For details on the weights see the [Accounting for Censoring in Performance Metrics for Event Time Data](../survival-metrics-details/) article.
:::

Of the metrics we calculated with these predictions, let's take a look at the AUC ROC first.

> 

Reviewer: I think that it more helpful to look at plots of the data for these two. Maybe something like

```r
collect_metrics(res_survreg) %>% 
  filter(.metric == "roc_auc_survival") %>% 
  ggplot(aes(.eval_time, mean)) + 
  geom_line() + 
  labs(x = "Evaluation Time", y = "Area Under the ROC Curve")

collect_metrics(res_survreg) %>% 
  filter(.metric == "brier_survival") %>% 
  ggplot(aes(.eval_time, mean)) + 
  geom_line() + 
  labs(x = "Evaluation Time", y = "Brier Score")
```
and initial show `collect_metrics` with a filter on "brier_survival_integrated" (to show the tabular results)

```{r}
collect_metrics(res_survreg) %>% 
  filter(.metric %in% c("brier_survival_integrated", "brier_survival"))
```

The accuracy of the predicted probabilities is generally good, albeit lowest for evaluation times of 30 and 60 days. The integrated Brier score is a measure of the overall accuracy of the predicted probabilities. 

Which metric to optimise for depends on whether separation or calibration is more important in the modeling problem at hand. We'll go with calibration here. Since we don't have a particular evaluation time that we want to predict well at, we are going to use the integrated Brier score as our main performance metric.

## Try out more models

Lumping factor levels together based on frequencies can lead to a loss of information so let's also try some different approaches. We can let a random forest model group the factor levels via the tree splits. Alternatively, we can turn the factors into dummy variables and use a regularized model to select relevant factor levels.

> First, we are creating the recipes for these two approaches.

Reviewer: ```suggestion
First, let’s create the recipes for these two approaches:
```

```{r}
rec_unknown <- recipe(disposition_surv ~ ., data = complaints_train) %>% 
  step_unknown(complaint_priority) 

rec_dummies <- rec_unknown %>% 
  step_novel(all_nominal_predictors()) %>%
  step_dummy(all_nominal_predictors()) %>% 
  step_zv(all_predictors()) %>% 
  step_normalize(all_numeric_predictors())
```

> Next, we are creating the model specifications and tagging several hyperparameters for tuning. 

Reviewer: ```suggestion
Next, let's create the model specifications and tag several hyperparameters for tuning. 
```

For the random forest, we are using the `"aorsf"` engine for accelerated oblique random survival forests. An oblique tree can split on linear combinations of the predictors, i.e., it provides more flexibility in the splits than a tree which splits on a single predictor.
For the regularized model, we are using the `"glmnet"` engine for a semi-parametric Cox proportional hazards model.

```{r}

> spec_oblique <- rand_forest(mtry = tune(), min_n = tune()) %>% 

Reviewer: In all of our other training materials, we use the object type as a _suffix_. It would be good to be consistent with those.

  set_engine("aorsf") %>% 
  set_mode("censored regression")

wflow_oblique <- workflow() %>% 
  add_recipe(rec_unknown) %>% 
  add_model(spec_oblique)

spec_coxnet <- proportional_hazards(penalty = tune()) %>% 
  set_engine("glmnet") %>% 
  set_mode("censored regression")

wflow_coxnet <- workflow() %>% 
  add_recipe(rec_dummies) %>% 
  add_model(spec_coxnet)
```

We can tune workflows with any of the `tune_*()` functions such as `tune_grid()` for grid search or `tune_bayes()` for Bayesian optimization. Here we are using grid search for simplicity.

```{r}
set.seed(1)
res_oblique <- tune_grid(
  wflow_oblique,
  complaints_rset,
  grid = 10,
  metrics = metric_set(brier_survival_integrated, brier_survival, roc_auc_survival, concordance_survival),

>   eval_time = seq(0, 300, 30), 

Reviewer: DRY: let's create objects for the eval times and metrics and reference those. 

Also, let's name arguments after the first one.

  control = control_grid(save_workflow = TRUE)
)

set.seed(1)
res_coxnet <- tune_grid(
  wflow_coxnet,
  complaints_rset,
  grid = 10,
  metrics = metric_set(brier_survival_integrated, brier_survival, roc_auc_survival, concordance_survival),
  eval_time = seq(0, 300, 30), 
  control = control_grid(save_workflow = TRUE)
)
```

So do any of these models perform better than the parametric survival model?

```{r}
show_best(res_oblique, metric = "brier_survival_integrated", n = 5)

show_best(res_coxnet, metric = "brier_survival_integrated", n = 5)
```

```{r}
#| echo: false
bi_best_cox <- show_best(res_coxnet, metric = "brier_survival_integrated", n = 1) %>% pull(mean)
bi_best_rf <- show_best(res_oblique, metric = "brier_survival_integrated", n = 1) %>% pull(mean)
bi_best_survreg <- show_best(res_survreg, metric = "brier_survival_integrated", n = 1) %>% pull(mean)

if (bi_best_cox <= bi_best_rf || bi_best_survreg <= bi_best_rf) {
  stop("The best model changed.")
}
```

> The best regularized Cox model performs a little better than the parametric survival model, with an integrated Brier score of `r round(bi_best_cox, 4)` compared to `r round(bi_best_survreg, 4)` for the parametric model. The random forest performs yet a little better with an integrated Brier score of `r round(bi_best_rf, 4)`.

Reviewer: ```suggestion
The best regularized Cox model performs a little better than the parametric survival model, with an integrated Brier score of `r round(bi_best_cox, 4)` compared to `r round(bi_best_survreg, 4)` for the parametric model. However, the regularized model has the advantage of using fewer predictors in the model. The random forest performs a little better with an integrated Brier score of `r round(bi_best_rf, 4)`.
```

Author: Do you mean that as an advantage of the regularized model compared to the parametric model or the random forest? The placement of the sentence together with "however" and the circumstance that the regularized model is already better than the parametric one, throws me a little here.

## The final model

We chose the random forest model as the final model. So let's finalize the workflow by replacing the `tune()` placeholders with the best hyperparameters.

```{r}
param_best <- select_best(res_oblique, metric = "brier_survival_integrated")

wflow_oblique_best <- finalize_workflow(
  wflow_oblique,
  param_best
)
```

We can now fit the final model on the training data and evaluate it on the test data.

```{r}
set.seed(2)
fit_oblique <- last_fit(
  wflow_oblique_best, 
  complaints_split,
  metrics = metric_set(brier_survival_integrated, brier_survival, roc_auc_survival, concordance_survival),
  eval_time = seq(0, 300, 30)
)

collect_metrics(fit_oblique) %>% 
  filter(.metric == "brier_survival_integrated")
```

> 

Reviewer: It's a bit of code but we can also show how similar the validation and test set results are: 

```r
collect_metrics(res_survreg) %>% 
  filter(.metric == "brier_survival") %>% 
  mutate(Data = "Validation") %>% 
  bind_rows(
    collect_metrics(fit_oblique) %>% 
      filter(.metric == "brier_survival") %>% 
      mutate(Data = "Testing") %>% 
      rename(mean = .estimate)
  ) %>% 
  ggplot(aes(.eval_time, mean, col = Data)) + 
  geom_line() + 
  labs(x = "Evaluation Time", y = "Brier Score")
```

![image](https://github.com/tidymodels/tidymodels.org/assets/5731043/f1560f55-29f3-4eb4-91fc-05f8ffb4a93d)

Author: I've changed this to use the Brier scores on the validation set by the oblique forest (rather than the parametric model).
